### DISCUSSION TITLE
URL has rewrite.cfm injected

### DISCUSSION THREAD

**Discussion by cfcoder** [wheels-dev/wheels#1311] [Q&A] [answered]
New to CFWheels here. Checked out into my local sandbox an app written in CFWheels that a fellow team mate is working on. It runs fine for him. But for me every url has rewrite.cfm inserted into it such as:
`http://127.0.0.1:62322/rewrite.cfm/projects/overview/764`

For him, he doesn't see that. There is a second problem that may be impacted by this rewrite.cfm:
All the <script> tags for JS references such as jQuery library files that are located in /vendors is not found. They all get 404 errors. I'm not sure but I suspect this rewrite.cfm issue is probably causing this secondary issue.

I've tried two browsers (Brave and Firefox) and both have the same issue for me. For him, no issues. It's supposedly identical code since it came from the same SVN repo as he is writing to.

Any ideas what is causing this insertion and why?  Why is it different for him?

I am using Commandbox to start up my CF server. He is not. This may explain some difference in behavior.
CFWheels ver is 2.4.0

Here's the console view:
![image](https://github.com/cfwheels/cfwheels/assets/4595721/15677e80-ed53-4d2c-9ebd-61e67ca07726)

**Comment by bpamiri** [comment]
Bruce (@cfcoder ) you are absolutely correct this is due to the URL rewriting setting of the CFEngine. There are lots of different ways to configure this and there is a guide page at https://guides.cfwheels.org/cfwheels-guides/handling-requests-with-controllers/url-rewriting that describes some of this. 

Since you said your team mate started this codebase and doesn't use CommandBox then he probably doesn't have the files CommandBox needs to know how to handle this. You need to add two files to the root of your application. Ideally you'll want to check that into SVN as well.

The first file is going to be a server.json file.
```
{
	"name":"|appName|",
	"force":true,
	"web":{
			"http":{
					"host":"localhost"
			},
			"rewrites":{
					"enable":true,
					"config":"urlrewrite.xml"
			}
	},
	"app":{
			"cfengine":"|cfmlEngine|"
	}
}
```
Change |appname| and |cfmlEngine| to your own values. You could use myApp and Lucee for example. The key element of this file is the rewrites setting. 

The second file you'll need to add is a url rewrite.xml file.
```
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE urlrewrite
    PUBLIC "-//tuckey.org//DTD UrlRewrite 4.0//EN"
    "http://www.tuckey.org/res/dtds/urlrewrite4.0.dtd">
<!--
This file is intended for use with Tuckey UrlRewriteFilter (http://tuckey.org/urlrewrite/) for Tomcat Servlet Containers
Tuckey is used by commandbox (https://www.ortussolutions.com/products/commandbox) which can greatly speed your local development.
In order to use this file with commandbox, we recommend creating a server.json file in your webroot with this starter markup:

{
    "name":"myWebApp",
    "force":true,
    "web":{
        "http":{
            "host":"localhost",
            "port":60001
        },
        "rewrites":{
            "enable":true,
            "config":"urlrewrite.xml"
        }
    }
}

See more at http://docs.cfwheels.org/docs/commandbox and https://ortus.gitbooks.io/commandbox-documentation/content/embedded_server/serverjson.html
If you do not use Tuckey or Commandbox, this file can be safely deleted.
-->
<urlrewrite>
  <rule enabled="true">
    <name>CFWheels pretty URLs</name>
    <condition type="request-uri" operator="notequal">^/(cf_script|flex2gateway|jrunscripts|cfide|cfformgateway|cffileservlet|lucee|files|images|javascripts|miscellaneous|stylesheets|wheels/public/assets|robots.txt|favicon.ico|sitemap.xml|rewrite.cfm)</condition>
    <from>^/(.*)$</from>
    <to type="passthrough">/rewrite.cfm/$1</to>
  </rule>
</urlrewrite>
```

Let me know if that doesn't work for you.

**Reply by cfcoder** [reply]
Thanks for responding. The code base I'm using has both files and set as you say. I'm puzzled that you suggest the rewrites/enable setting is true. However, I tried it true and false and neither way made any difference, it still injects the rewrite.cfm. I don't want it there in the URL.  So I'm not sure how your suggestion solves my issue.

**Reply by cfcoder** [reply]
I set enabled to false and **restarted** (previously thought I had after setting enabled to false, but I hadn't) the CF server in commandbox. That makes the problem go away. JQuery file references in the layout.cfm file now resolve without 404 errors. So apparently Commandbox doesn't need the rewrite in the URL. Not sure why your docs says I need it.

I do see index.cfm in the URL where the rewrite.cfm used to be. So I guess if I want "pretty" URLs without index I need to turn rewriting back on. But I'm not sure how to then prevent the rewrite.cfm from being used?

**Comment by bpamiri** [comment]
Can you post your rewrite.xml file here. It looks like you are able to turn URL rewriting on and off but perhaps the rewrite.xml file in not being interpreted correctly to tell the server to pass everything through rewrite.cfm without actually changing the URL to it.

**Reply by cfcoder** [reply]
Here's my urlrewrite.xml file content:

```
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE urlrewrite
    PUBLIC "-//tuckey.org//DTD UrlRewrite 4.0//EN"
    "http://www.tuckey.org/res/dtds/urlrewrite4.0.dtd">
<!--
This file is intended for use with Tuckey UrlRewriteFilter (http://tuckey.org/urlrewrite/) for Tomcat Servlet Containers
Tuckey is used by commandbox (https://www.ortussolutions.com/products/commandbox) which can greatly speed your local development.
In order to use this file with commandbox, we recommend creating a server.json file in your webroot with this starter markup:

{
    "name":"myWebApp",
    "force":true,
    "web":{
        "http":{
            "host":"localhost",
            "port":60001
        },
        "rewrites":{
            "enable":true,
            "config":"urlrewrite.xml"
        }
    }
}

See more at http://docs.cfwheels.org/docs/commandbox and https://ortus.gitbooks.io/commandbox-documentation/content/embedded_server/serverjson.html
If you do not use Tuckey or Commandbox, this file can be safely deleted.
-->
<urlrewrite>
  <rule enabled="true">
    <name>CFWheels pretty URLs</name>
    <condition type="request-uri" operator="notequal">^/(cf_script|flex2gateway|jrunscripts|cfide|cfformgateway|cffileservlet|lucee|files|images|javascripts|miscellaneous|stylesheets|wheels/public/assets|robots.txt|favicon.ico|sitemap.xml|rewrite.cfm)</condition>
    <from>^/(.*)$</from>
    <to type="passthrough">/rewrite.cfm/$1</to>
  </rule>
</urlrewrite>
```

**Comment by bpamiri** [comment]
I'm stumped but it's probably going to be a simple thing we are overlooking. So let's go back to your original URL you posted:
```http://127.0.0.1:62322/rewrite.cfm/projects/overview/764```
What happens if you issue:
```http://127.0.0.1:62322/projects/overview/764```

I'm trying to determine if the URL Rewriting is working fine but the urlFor() function is inserting the rewrite.cfm into the URL.

**Reply by cfcoder** [reply]
It loads the same page, but puts the rewrite.cfm in the path again. I think it is that file content above, the "to" tag is inserting that "/rewrite.cfm/$1".  I just don't know how that is making the URL "pretty" as the docs say.  To me it's adding this to the path and breaking things. It essentially changes index.cfm to rewrite.cfm in the url.

**Reply by MvdO79** [reply]
It shouldn't insert "/rewrite.cfm/$1". So something is indeed not working. Just to be sure, can you show your server.json?

**Reply by bpamiri** [reply]
This is a typical server.json that I get when I create a new app using the CommandBox wheels command. 
```
{
    "name":"myapp",
    "force":true,
    "web":{
        "http":{
            "host":"localhost"
        },
        "rewrites":{
            "enable":true,
            "config":"urlrewrite.xml"
        }
    },
    "app":{
        "cfengine":"lucee"
    }
}
```

**Reply by bpamiri** [reply]
and the url rewrite.xml file.
```
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE urlrewrite
    PUBLIC "-//tuckey.org//DTD UrlRewrite 4.0//EN"
    "http://www.tuckey.org/res/dtds/urlrewrite4.0.dtd">
<!--
This file is intended for use with Tuckey UrlRewriteFilter (http://tuckey.org/urlrewrite/) for Tomcat Servlet Containers
Tuckey is used by commandbox (https://www.ortussolutions.com/products/commandbox) which can greatly speed your local development.
In order to use this file with commandbox, we recommend creating a server.json file in your webroot with this starter markup:

{
    "name":"myWebApp",
    "force":true,
    "web":{
        "http":{
            "host":"localhost",
            "port":60001
        },
        "rewrites":{
            "enable":true,
            "config":"urlrewrite.xml"
        }
    }
}

See more at http://docs.cfwheels.org/docs/commandbox and https://ortus.gitbooks.io/commandbox-documentation/content/embedded_server/serverjson.html
If you do not use Tuckey or Commandbox, this file can be safely deleted.
-->
<urlrewrite>
  <rule enabled="true">
    <name>CFWheels pretty URLs</name>
    <condition type="request-uri" operator="notequal">^/(cf_script|flex2gateway|jrunscripts|cfide|cfformgateway|cffileservlet|lucee|files|images|javascripts|miscellaneous|stylesheets|wheels/public/assets|robots.txt|favicon.ico|sitemap.xml|rewrite.cfm)</condition>
    <from>^/(.*)$</from>
    <to type="passthrough">/rewrite.cfm/$1</to>
  </rule>
</urlrewrite>
```

**Reply by bpamiri** [reply]
the last thing to check is the URLRewriting setting in config/settings.cfm
```
	/*
		If you comment out the following line, CFWheels will try to determine the URL rewrite capabilities automatically.
		The "URLRewriting" setting can bet set to "on", "partial" or "off".
		To run with "partial" rewriting, the "cgi.path_info" variable needs to be supported by the web server.
		To run with rewriting set to "on", you need to apply the necessary rewrite rules on the web server first.
	*/
	set(URLRewriting="On");
```

**Reply by bpamiri** [reply]
The framework tries to guess your URL rewrite setup and for the most part it guesses fine but in some instances you have to force it into a particular mode.

I always specify it as that is what I usually want.

**Comment by cfcoder** [comment]
My server.json file is what bpamiri posted above except I'm using cfengine adobe@2018.
My urlrewrite.cml is what bpamiri posted above as well. I think the injection of rewrite.cfm is happening in that file rules:
```
<from>^/(.*)$</from>
<to type="passthrough">/rewrite.cfm/$1</to>

```
It's clearly instructing to add rewrite.cfm into the url. But that file is provided by CFW install. I did use Commandbox to install. So perhaps that auto-detect is failing and generating the wrong rule?

As for the config/setting.cfm file, my relevant command is different from bpamiri. I have:
set(URLRewriting="partial");

**Reply by bpamiri** [reply]
That is why the framework is injecting rewrite.cfm into the URL. Change the setting to on, then reload the app and see if that clears that up.

**Comment by cfcoder** [comment]
Ah-ha! That change works. I get the "pretty" version of the URL without any "rewrite.cfm" insertion. Thanks for that solution!

**Comment by cfcoder** [comment]
I am having a new related issue. In the project tree root is a "/vendors" directory where all the jQuery files live. With set(URLRewriting="On"); setting set to "on" (from "partial") and the rewrite.xml file containing the enable = true, the /vendors directory files are not being found. It is most strange. If I enable to false, then it works but insert "index.cfm" into the url path and the "pretty" version of each path is gone. What is going on?? The files do exist at the path indicated in the screenshot, but CFWheels doesn't think so.

```
 "rewrites":{
            "enable":true,
            "config":"urlrewrite.xml"
        }

```
![image](https://github.com/cfwheels/cfwheels/assets/4595721/a3a43c0e-1c17-4cf7-8e25-19506561422a)

**Reply by cfcoder** [reply]
I should add that in the head tag are these script and link tags, all are failing. However, those that use the javaScriptIncludeTag() function are fine. But then those resource files are located in a different directory (/scripts).
```
<script src="/vendors/jquery/jquery.js"></script>
<link rel="stylesheet" href="/vendors/jquery-ui-1.13.2/jquery-ui.min.css">
<script src="/vendors/jquery-ui-1.13.2/jquery-ui.js"></script>
<link rel="stylesheet" href="/vendors/fontawesome-free-5.15.4-web/css/all.css">
```

File paths exist in the source.
![image](https://github.com/cfwheels/cfwheels/assets/4595721/73540b49-2db4-4b82-a93a-689570d3b62d)

**Comment by bpamiri** [comment, marked as answer]
This is by design and blocks access to directories on your server with out going through the framework. But to break out of it you need to modify the urlrewrite.xml file. Take a look at the following line of code:

https://github.com/cfwheels/cfwheels/blob/d9cdbea3e64177d0b6b050780c2cf22dbb2da9ed/urlrewrite.xml#L31

Add "vendor/jquery" to the list of URL prefixes to ignore. These items are "|" delimited so make sure to add that after or before depending where in the list you add your "vendor/jquery" entry.

Then restart your server via `box server restart` and that should solve your issue.

**Reply by cfcoder** [reply]
I was able to add "vendor" to the list and that appears to solve the problem.

Since our test and production servers are not controlled by Commandbox,  how is the configuration settings in the server.json file added so that the rewrites are enabled and configured to use the urlrewrite.xml in CFW projects? Or does CFWheels read that server.json file, too?

**Reply by bpamiri** [reply]
Bruce,

URL Rewriting, or better said, shortening the URL in order to use "pretty" URLs or search engine safe URLs is a function of the web server, application server, and OS you use. The included server.json and URLRewrite.xml files are include to make development with CommandBox simpler. But how your staging server or production server are configure will determine how you have to configure the URL rewriting rules. For example, if you use Tomcat, Jetty, or another equivalent JVM servlet engine to server the site directly you will most likely need to setup the application server to use URLRewrite.xml but if you front that with an Apache server, then your URL rewriting is happening in Apache and you'll have to configure an .htaccess file with the rules in there.

What your are doing by adding `vendor` is augmenting the default set of rules by adding another folder to the list. Spoiler alert, a future version of Wheels may have this setting preconfigured. ;-)

Please take a look at https://guides.cfwheels.org/cfwheels-guides/handling-requests-with-controllers/url-rewriting to see the documentation on how to setup different environments.

**Reply by cfcoder** [reply]
Thanks @bpamiri !
